cmd/apiserver/server_test.go

```diff
@@ -0,0 +1,401 @@
+
+	teff := 10000
+	testCalculationCount := teff + 2
+	var calcList []*calculationsv1.Calculation
+	for teff != testCalculationCount {
+		teff++
+		newCalc := util.NewCalculation(float64(teff), 4.0)
+
+		calcList = append(calcList, newCalc)
+		logrus.WithFields(logrus.Fields{"calculation": newCalc.Name, "teff": teff, "logG": "4.0"}).Info("Creating calculation for test purpose")
+		_, err := fakeClient.Calculations().Create(ctx, newCalc, metav1.CreateOptions{})
+		if err != nil {
+			t.Errorf("Coudn't create the calculation: %v", newCalc)
+		} else {
+			logrus.WithFields(logrus.Fields{"calculation": newCalc.Name, "teff": newCalc.Spec.Teff, "logG": newCalc.Spec.LogG}).Info("Calculation created successfully")
+		}
+	}
+
+	s := server{
+		logger: logrus.WithField("test-name", "get calculations test"),
+		ctx:    ctx,
+		client: fakeClient,
+	}
+
+	req, err := http.NewRequest("DELETE", fmt.Sprintf("/calculations/delete/%v", calcList[0].Name), nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	q := req.URL.Query()
+	q.Add("name", calcList[0].Name)
+	req.URL.RawQuery = q.Encode()
+
+	req.Header.Set("Content-Type", "application/json")
+	rr := httptest.NewRecorder()
+	handler := http.HandlerFunc(s.deleteCalculation)
+	handler.ServeHTTP(rr, req)
+
+	response := Response{}
+	err = json.Unmarshal(rr.Body.Bytes(), &response)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if status := response.StatusCode; status != http.StatusOK && status != 0 {
+		logrus.Infof("Handler returned wrong status code: got %v want %v", status, http.StatusOK)
+	}
+
+	var actualData *calculationsv1.CalculationList
+
+	err = json.Unmarshal(rr.Body.Bytes(), &actualData)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if diff := cmp.Diff(calcList[1], actualData.Items); diff != "" {
+		t.Fatal(diff)
+	}
+}
+
+func TestGetCalculations(t *testing.T) {
+	testCases := []struct {
+		id               string
+		calculationCount int
+		expected         []calculationsv1.Calculation
+	}{
+		{
+			id:               "get no calculation",
+			calculationCount: 0,
+		},
+		{
+			id:               "get two calculations",
+			calculationCount: 2,
+		},
+	}
+
+	fakecs := fake.NewSimpleClientset()
+	var fakeClient v1.VegaV1Interface
+	fakeClient = fakecs.VegaV1()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	teff := 10000
+
+	for _, tc := range testCases {
+		var calculationCount = teff + tc.calculationCount
+		for teff != calculationCount {
+			teff++
+			newCalculation := util.NewCalculation(float64(teff), 4.0)
+			logrus.WithFields(logrus.Fields{"teff": teff, "logG": "4.0"}).Info("Creating calculation for test purpose")
+			_, err := fakeClient.Calculations().Create(ctx, newCalculation, metav1.CreateOptions{})
+			if err != nil {
+				t.Errorf("Coudn't create the calculation: %v", newCalculation)
+			} else {
+				logrus.WithFields(logrus.Fields{"teff": teff, "logG": "4.0"}).Info("Calculation successfully created")
+			}
+			tc.expected = append(tc.expected, *newCalculation)
```

Comment:
every testcase should hold a different set of calculations.

Reply:
But that is what I am doing already. I am iterating `teff`. So there are calculations with different teff values, `logG` is same. I dont change that.

Reply:
Every testcase should hold the initial calculation and then you can create them.

Reply:
`func TestGetCalculations(t *testing.T) {
	type values struct {
		teff float64
		logG float64
	}

	testCases := []struct {
		id                string
		calculationValues []values
		expected          []calculationsv1.Calculation
	}{
		{
			id: "get no calculation",
		},
		{
			id: "get one calculations",
			calculationValues: []values{
				{
					teff: 10000.0,
					logG: 4.0,
				},
			},
		},
		{
			id: "get two calculations",
			calculationValues: []values{
				{
					teff: 11000.0,
					logG: 4.0,
				},
				{
					teff: 12000.0,
					logG: 4.0,
				},
			},
		},
	}

	fakecs := fake.NewSimpleClientset()
	fakeClient := fakecs.VegaV1()

	for _, tc := range testCases {
		for _, value := range tc.calculationValues {
			newCalculation := util.NewCalculation(value.teff, value.logG)
			logrus.WithFields(logrus.Fields{"teff": value.teff, "logG": value.logG}).Info("Creating calculation for test purpose")
			_, err := fakeClient.Calculations().Create(context.Background(), newCalculation, metav1.CreateOptions{})
			if err != nil {
				t.Errorf("Coudn't create the calculation: %v", newCalculation)
			} else {
				logrus.WithFields(logrus.Fields{"teff": value.teff, "logG": value.logG}).Info("Calculation successfully created")
			}
			tc.expected = append(tc.expected, *newCalculation)
		}

		s := server{
			logger: logrus.WithField("test-name", tc.id),
			ctx:    context.Background(),
			client: fakeClient,
		}

		req, err := http.NewRequest("GET", "/calculations", nil)
		if err != nil {
			t.Fatal(err)
		}
		req.Header.Set("Content-Type", "application/json")
		rr := httptest.NewRecorder()
		handler := http.HandlerFunc(s.getCalculations)
		handler.ServeHTTP(rr, req)
		if status := rr.Code; status != http.StatusOK {
			t.Errorf("Handler returned wrong status code: got %v want %v",
				status, http.StatusOK)
		}

		var actualData *calculationsv1.CalculationList

		err = json.Unmarshal(rr.Body.Bytes(), &actualData)
		if err != nil {
			t.Fatal(err)
		}

		if diff := cmp.Diff(tc.expected, actualData.Items); diff != "" {
			t.Fatal(diff)
		}
	}
}`

Reply:
`
func TestGetCalculations(t *testing.T) {
	type values struct {
		teff float64
		logG float64
	}

	testCases := []struct {
		id                string
		calculationValues []values
		expected          []calculationsv1.Calculation
	}{
		{
			id: "get no calculation",
		},
		{
			id: "get one calculations",
			calculationValues: []values{
				{
					teff: 10000.0,
					logG: 4.0,
				},
			},
		},
		{
			id: "get two calculations",
			calculationValues: []values{
				{
					teff: 11000.0,
					logG: 4.0,
				},
				{
					teff: 12000.0,
					logG: 4.0,
				},
			},
		},
	}

	fakecs := fake.NewSimpleClientset()
	fakeClient := fakecs.VegaV1()

	for _, tc := range testCases {
		for _, value := range tc.calculationValues {
			newCalculation := util.NewCalculation(value.teff, value.logG)
			logrus.WithFields(logrus.Fields{"teff": value.teff, "logG": value.logG}).Info("Creating calculation for test purpose")
			_, err := fakeClient.Calculations().Create(context.Background(), newCalculation, metav1.CreateOptions{})
			if err != nil {
				t.Errorf("Coudn't create the calculation: %v", newCalculation)
			} else {
				logrus.WithFields(logrus.Fields{"teff": value.teff, "logG": value.logG}).Info("Calculation successfully created")
			}
			tc.expected = append(tc.expected, *newCalculation)
		}

		s := server{
			logger: logrus.WithField("test-name", tc.id),
			ctx:    context.Background(),
			client: fakeClient,
		}

		req, err := http.NewRequest("GET", "/calculations", nil)
		if err != nil {
			t.Fatal(err)
		}
		req.Header.Set("Content-Type", "application/json")
		rr := httptest.NewRecorder()
		handler := http.HandlerFunc(s.getCalculations)
		handler.ServeHTTP(rr, req)
		if status := rr.Code; status != http.StatusOK {
			t.Errorf("Handler returned wrong status code: got %v want %v",
				status, http.StatusOK)
		}

		var actualData *calculationsv1.CalculationList

		err = json.Unmarshal(rr.Body.Bytes(), &actualData)
		if err != nil {
			t.Fatal(err)
		}

		if diff := cmp.Diff(tc.expected, actualData.Items); diff != "" {
			t.Fatal(diff)
		}
	}
}
`

Reply:
`func TestGetCalculations(t *testing.T) {
	
        type values struct {
		teff float64
		logG float64
	}

	testCases := []struct {
		id                string
		calculationValues []values
		expected          []calculationsv1.Calculation
	}{
		{
			id: "get no calculation",
		},
		{
			id: "get one calculations",
			calculationValues: []values{
				{
					teff: 10000.0,
					logG: 4.0,
				},
			},
		},
		{
			id: "get two calculations",
			calculationValues: []values{
				{
					teff: 11000.0,
					logG: 4.0,
				},
				{
					teff: 12000.0,
					logG: 4.0,
				},
			},
		},
	}

	fakecs := fake.NewSimpleClientset()
	fakeClient := fakecs.VegaV1()

	for _, tc := range testCases {
		for _, value := range tc.calculationValues {
			newCalculation := util.NewCalculation(value.teff, value.logG)
			logrus.WithFields(logrus.Fields{"teff": value.teff, "logG": value.logG}).Info("Creating calculation for test purpose")
			_, err := fakeClient.Calculations().Create(context.Background(), newCalculation, metav1.CreateOptions{})
			if err != nil {
				t.Errorf("Coudn't create the calculation: %v", newCalculation)
			} else {
				logrus.WithFields(logrus.Fields{"teff": value.teff, "logG": value.logG}).Info("Calculation successfully created")
			}
			tc.expected = append(tc.expected, *newCalculation)
		}

		s := server{
			logger: logrus.WithField("test-name", tc.id),
			ctx:    context.Background(),
			client: fakeClient,
		}

		req, err := http.NewRequest("GET", "/calculations", nil)
		if err != nil {
			t.Fatal(err)
		}
		req.Header.Set("Content-Type", "application/json")
		rr := httptest.NewRecorder()
		handler := http.HandlerFunc(s.getCalculations)
		handler.ServeHTTP(rr, req)
		if status := rr.Code; status != http.StatusOK {
			t.Errorf("Handler returned wrong status code: got %v want %v",
				status, http.StatusOK)
		}

		var actualData *calculationsv1.CalculationList

		err = json.Unmarshal(rr.Body.Bytes(), &actualData)
		if err != nil {
			t.Fatal(err)
		}

		if diff := cmp.Diff(tc.expected, actualData.Items); diff != "" {
			t.Fatal(diff)
		}
	}
}`

Reply:
Those should be the changes to the function. At least I hope so. But I am still getting diff at the end.

`        - 			StartTime:      v1.Time{Time: s"2021-04-08 14:59:24.554503144 +0200 CEST m=+0.003896033"},
        + 			StartTime:      v1.Time{Time: s"2021-04-08 14:59:24 +0200 CEST"},
`

Reply:
Well, the whole concept here is wrong. The testcase holds only the teff,logG values and then you use `util.NewCalculation` to create them. Instead, we need the testcase to hold the calculation structs itself and then compare them with the expected ones.

Reply:
But I thought the expected calculations are the ones I get from `util.NewCalculation` and then the actual calculations I get from JSON.

Reply:
The `util.NewCalculation` returns a `Calculation`. You can initialize this in the testcase with the values you need to compare.

Reply:
So I should have in testCases something like this - `newCalculation: util.NewCalculation(10000.0, 4.0)` and then in the loop through all test cases have `_, err := fakeClient.Calculations().Create(context.Background(), tc.newCalculation, metav1.CreateOptions{})`. I am sorry, but I dont get why isn't the code I had correct :laughing:

Reply:
You should have a list of calculations in the test case that you will create and the same expected ones and then you can compare.

Reply:
But I thought that ones that I create are expected and actual are the ones I get from the api server.

Reply:
The ones that you get from the API server are the ones that you will create first.  Something like:
```go
[]v1.Calculation{
	{
		TypeMeta:   metav1.TypeMeta{Kind: "Calculation", APIVersion: "vega.io/v1"},
		ObjectMeta: metav1.ObjectMeta{Name: "calc-1"},
		Phase:      v1.CreatedPhase,
		Status:     v1.CalculationStatus{StartTime: metav1.Time{Time: time.Date(2000, 0, 0, 0, 0, 0, 0, time.UTC)}},
		Spec:       v1.CalculationSpec{Teff: 1000.0 ,LogG: 4.0}
	},
	{
		TypeMeta:   metav1.TypeMeta{Kind: "Calculation", APIVersion: "vega.io/v1"},
		ObjectMeta: metav1.ObjectMeta{Name: "calc-2"},
		Phase:      v1.CreatedPhase,
		Status:     v1.CalculationStatus{StartTime: metav1.Time{Time: time.Date(2000, 0, 0, 0, 0, 0, 1, time.UTC)}},
		Spec:       v1.CalculationSpec{Teff: 1100.0 ,LogG: 4.0}
	},
}
```

Reply:
So I create calculation like this. Then put it into `expected`. And compare these to what I got from api server(what I unmarshal). And that's it, right?

Reply:
```go
testCases := []struct {
		id                   string
		initialCalculations  []Calculation
		expectedCalculations []Calculation
	}{
		{
			id: "one calculation case",
			initialCalculations: []v1.Calculation{
				{
					TypeMeta:   metav1.TypeMeta{Kind: "Calculation", APIVersion: "vega.io/v1"},
					ObjectMeta: metav1.ObjectMeta{Name: "calc-1"},
					Phase:      v1.CreatedPhase,
					Status:     v1.CalculationStatus{StartTime: metav1.Time{Time: time.Date(2000, 0, 0, 0, 0, 0, 0, time.UTC)}},
					Spec:       v1.CalculationSpec{Teff: 1000.0, LogG: 4.0},
				},
			},
			expectedCalculations: []v1.Calculation{
				{
					TypeMeta:   metav1.TypeMeta{Kind: "Calculation", APIVersion: "vega.io/v1"},
					ObjectMeta: metav1.ObjectMeta{Name: "calc-1"},
					Phase:      v1.CreatedPhase,
					Status:     v1.CalculationStatus{StartTime: metav1.Time{Time: time.Date(2000, 0, 0, 0, 0, 0, 0, time.UTC)}},
					Spec:       v1.CalculationSpec{Teff: 1000.0, LogG: 4.0},
				},
			},
		},
.......
.....
.....
	}

```

Every test case will create the `tc.initialCalculations` , and then you will compare what you got with the `tc.expectedCalculations`